drivers/usb/udc/udc_dwc2_vendor_quirks.h

```diff
@@ -96,6 +108,158 @@ DT_INST_FOREACH_STATUS_OKAY(QUIRK_STM32F4_FSOTG_DEFINE)
 
 #endif /*DT_HAS_COMPAT_STATUS_OKAY(st_stm32f4_fsotg) */
 
+#if DT_HAS_COMPAT_STATUS_OKAY(nordic_nrf_usbhs)
+
+#define DT_DRV_COMPAT snps_dwc2
+
+#include <nrfs_backend_ipc_service.h>
+#include <nrfs_usb.h>
+
+#define USBHS_DT_WRAPPER_REG_ADDR(n) UINT_TO_POINTER(DT_INST_REG_ADDR_BY_NAME(n, wrapper))
+
+/*
+ * On USBHS, we cannot access the DWC2 register until VBUS is detected and
+ * valid. If the user tries to force usbd_enable() and the corresponding
+ * udc_enable() without a "VBUS ready" notification, the event wait will block
+ * until a valid VBUS signal is detected.
+ */
+static K_EVENT_DEFINE(usbhs_events);
+#define USBHS_VBUS_READY	BIT(0)
+
+static void usbhs_vbus_handler(nrfs_usb_evt_t const *p_evt, void *const context)
+{
+	const struct device *dev = context;
+
+	switch (p_evt->type) {
+	case NRFS_USB_EVT_VBUS_STATUS_CHANGE:
+		LOG_DBG("USBHS new status, pll_ok = %d vreg_ok = %d vbus_detected = %d",
+			p_evt->usbhspll_ok, p_evt->vregusb_ok, p_evt->vbus_detected);
+
+		if (p_evt->usbhspll_ok && p_evt->vregusb_ok && p_evt->vbus_detected) {
+			k_event_post(&usbhs_events, USBHS_VBUS_READY);
+			udc_submit_event(dev, UDC_EVT_VBUS_READY, 0);
+		} else {
+			k_event_set_masked(&usbhs_events, 0, USBHS_VBUS_READY);
+			udc_submit_event(dev, UDC_EVT_VBUS_REMOVED, 0);
+		}
+
+		break;
+	case NRFS_USB_EVT_REJECT:
+		LOG_ERR("Request rejected");
+		break;
+	default:
+		LOG_ERR("Unknown event type 0x%x", p_evt->type);
+		break;
+	}
+}
+
+static inline int usbhs_enable_nrfs_service(const struct device *dev)
+{
+	nrfs_err_t nrfs_err;
+	int err;
+
+	err = nrfs_backend_wait_for_connection(K_MSEC(1000));
+	if (err) {
+		LOG_INF("NRFS backend connection timeout");
+		return err;
+	}
+
+	nrfs_err = nrfs_usb_init(usbhs_vbus_handler);
+	if (nrfs_err != NRFS_SUCCESS) {
+		LOG_ERR("Failed to init NRFS VBUS handler: %d", nrfs_err);
+		return -EIO;
+	}
+
+	nrfs_err = nrfs_usb_enable_request((void *)dev);
+	if (nrfs_err != NRFS_SUCCESS) {
+		LOG_ERR("Failed to enable NRFS VBUS service: %d", nrfs_err);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static inline int usbhs_enable_core(const struct device *dev)
+{
+	NRF_USBHS_Type *wrapper = USBHS_DT_WRAPPER_REG_ADDR(0);
+
+	if (!k_event_wait(&usbhs_events, USBHS_VBUS_READY, false, K_NO_WAIT)) {
+		LOG_WRN("VBUS is not ready, block udc_enable()");
+		k_event_wait(&usbhs_events, USBHS_VBUS_READY, false, K_FOREVER);
+	}
+
+	wrapper->ENABLE = USBHS_ENABLE_PHY_Msk | USBHS_ENABLE_CORE_Msk;
+	wrapper->TASKS_START = 1UL;
+
+	/* Enable interrupts */
+	wrapper->INTENSET = 1UL;
+
+	return 0;
+}
+
+static inline int usbhs_disable_core(const struct device *dev)
+{
+	NRF_USBHS_Type *wrapper = USBHS_DT_WRAPPER_REG_ADDR(0);
+
+	/* Disable interrupts */
+	wrapper->INTENSET = 0UL;
```

Comment:
This is not disabling interrupts. It is essentially no action.

Either change it to `wrapper->INTEN = 0UL;` or `wrapper->INTENCLR = 1UL;`.